### PR TITLE
Adds custom Nullable Annotation via Error Prone CLI flags

### PR DESCRIPTION
**Pull request by nimakarimipour**
This PR adds the ability to add custom ```@Nonnull / @Nullable``` annotation to the analysis.

I observed projects that are partially annotated with annotations other than conventional ```@Nullable``` annotations (e.g. ```javax.annotation.Nullable```). If we run Nullaway in these projects, the set of existing ```@Nullable``` annotations are not sufficient for NullAway to pass and needs further ```@Nullable``` injections. 

Since these annotations currently exist in the project, they have been verified by the programmers and can be integral part of the analysis if understood by NullAway. This PR adds the feature to introduce new ```@Nullable``` annotations to NullAway.

Please find project [LibGdx](https://github.com/libgdx/libgdx) where it has been partially annotated with [@com.badlogic.gdx.utils.Null](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/utils/Null.java) annotation using a custom ```@Nullable``` annotation declared in the project itself.

With best regards

### CODE REVIEW

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java [hunk 1]
```diff
@@ -3170,4 +3170,31 @@ public void derefNullableTernary() {
...
+            "     passNullable(foo);",
+            "     return null;", // No error, should detect @Null
+            "   }",
+            "   String passNullable(@Null Object item){",
+            "    // BUG: Diagnostic contains: dereferenced expression",
+            "      return item.toString();",
+            "   }",
+            "}")
+        .doTest();
+  }
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java:3199]: This would need, at the very least, also a test case for the non-null case as well, note that for that you probably need a package in the unannotated set and the `AcknowledgeRestrictiveAnnotations` option.

File: nullaway/src/main/java/com/uber/nullaway/ErrorProneCLIFlagsConfig.java [hunk 2]
```diff
@@ -152,6 +154,8 @@
             flags, FL_CLASS_ANNOTATIONS_TO_EXCLUDE, DEFAULT_CLASS_ANNOTATIONS_TO_EXCLUDE);
     initializerAnnotations =
         getFlagStringSet(flags, FL_INITIALIZER_ANNOT, DEFAULT_INITIALIZER_ANNOT);
+    customNullableAnnotation = flags.get(FL_NULLABLE_ANNOT).orElse("javax.annotation.Nullable");
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/ErrorProneCLIFlagsConfig.java:157]: We already support `javax.annotation.Nullable` (and anything with simple name `@Nullable` natively), so this default is redundant. Also, users might want to specify more than one custom nullability annotation at the time in a large repo. 

My suggestion here would be to model **both** of these new options after `CustomInitializerAnnotations`, using `getFlagStringSet`, but use an empty immutable set (`ImmutableSet.of()` without arguments) as the default.

Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/ErrorProneCLIFlagsConfig.java]: 1c56af6a5e5423c982c620cbd2476bdee23ee212

File: nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java [hunk 3]
```diff
@@ -110,6 +110,11 @@
 
   protected String errorURL;
 
+  /** --- Fully qualified names of custom nonnull/nullable annotation --- */
+  protected Set<String> customNonnullAnnotation;
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java:114]: These two should probably also be plural (annotation**s**), as should the comment above.

**Comment by msridhar** [comment]
Haven't gotten to review yet, but I want to note that checking for `@Nullable` annotations happens _a lot_ in NullAway.  I would hope to see that running `./gradlew jmh` before and after this change results in little to no perf regression in the case where custom annotations are not specified.  We may need to uglify the code a bit to get the perf overhead down.

**Comment by lazaroclapp** [comment]
> Haven't gotten to review yet, but I want to note that checking for `@Nullable` annotations happens _a lot_ in NullAway. I would hope to see that running `./gradlew jmh` before and after this change results in little to no perf regression in the case where custom annotations are not specified. We may need to uglify the code a bit to get the perf overhead down.

Agreed! That said, this adds an `ImmutableSet.contains` call in that critical code, which should be pretty fast. Still, definitely worth running performance testing before landing this PR (I can do that once the extra unit test case is there)

File: nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java (hunk 3 above)
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java]: 441dd40ea4d83e8b36e341fffc785b60a5e3b0ad

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 1 above)
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java:3212]: 88f220fc1bf75db4ca58f354be29a11488ef00dd

To my understanding, ```NullAway``` only looks for  explicit ```@Nonnull``` annotation on method returns for checking inheritance conditions. This test case targets this property. Please let me know if I should add more cases. Thank you.

**Comment by nimakarimipour** [comment]
> Haven't gotten to review yet, but I want to note that checking for `@Nullable` annotations happens _a lot_ in NullAway. I would hope to see that running `./gradlew jmh` before and after this change results in little to no perf regression in the case where custom annotations are not specified. We may need to uglify the code a bit to get the perf overhead down.

Hi @msridhar, Would you please let me know how I can test this property so I can test it myself and improve it if needed ? Thank you very much.

**Comment by msridhar** [comment]
> you please let me know how I can test this property so I can test it myself and improve it if needed ? Thank you very much.

@nimakarimipour you can just run `./gradlew jmh` for the master branch and for your PR branch, and compare the results.  You need to run this on JDK 11 or above, and each run takes ~25 minutes.  You should also try to quit other applications and background services as much as possible to reduce noise.  At the end you'll see a table like this summarizing the results:

```
Benchmark                      Mode  Cnt   Score   Error  Units
AutodisposeBenchmark.compile  thrpt   25  17.580 ± 0.143  ops/s
CaffeineBenchmark.compile     thrpt   25   4.196 ± 0.012  ops/s
```

**Comment by nimakarimipour** [comment]
> > you please let me know how I can test this property so I can test it myself and improve it if needed ? Thank you very much.
> 
> @nimakarimipour you can just run `./gradlew jmh` for the master branch and for your PR branch, and compare the results. You need to run this on JDK 11 or above, and each run takes ~25 minutes. You should also try to quit other applications and background services as much as possible to reduce noise. At the end you'll see a table like this summarizing the results:
> 
> ```
> Benchmark                      Mode  Cnt   Score   Error  Units
> AutodisposeBenchmark.compile  thrpt   25  17.580 ± 0.143  ops/s
> CaffeineBenchmark.compile     thrpt   25   4.196 ± 0.012  ops/s
> ```

Thank you very much for the instructions, please find the results below:

On ```master``` branch:
```
Benchmark                      Mode  Cnt   Score   Error  Units
AutodisposeBenchmark.compile  thrpt   25  11.500 ± 0.054  ops/s
CaffeineBenchmark.compile     thrpt   25   2.504 ± 0.021  ops/s
```

On ```custom_nullable``` branch (this PR):
```
Benchmark                      Mode  Cnt   Score   Error  Units
AutodisposeBenchmark.compile  thrpt   25  11.552 ± 0.075  ops/s
CaffeineBenchmark.compile     thrpt   25   2.560 ± 0.012  ops/s
```

Please let me know if I should do anything else for comparing performances.

**Comment by msridhar** [comment]
@nimakarimipour those numbers look good.  At least for our jmh benchmarks, it looks like the perf difference is in the noise.  I wonder if in the steady state, the Hotspot JIT is smart enough to completely simplify away the check (make it zero cost) when there are no custom annotations specified.  Would be fun to dig into 🙂 But not necessary to land this PR

**Comment by nimakarimipour** [comment]
> @nimakarimipour those numbers look good. At least for our jmh benchmarks, it looks like the perf difference is in the noise. I wonder if in the steady state, the Hotspot JIT is smart enough to completely simplify away the check (make it zero cost) when there are no custom annotations specified. Would be fun to dig into 🙂 But not necessary to land this PR

That's great. Regarding noise, would you like me to run this experiment in google cloud via a docker or that is noisy as well ?

**Comment by msridhar** [comment]
Don't worry about it.  You can't reduce noise to 0.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 1 above)
Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java:3212]: Hey, I think this makes sense to me. As a sanity check, I'd try to comment the line:

```
"-XepOpt:NullAway:CustomNonnullAnnotations=qual.NoNull",
```

and make sure the test fails then.

That said, another thing I am going to ask you to do here is to separate this into two unit test methods, one for `testCustomNullableAnnotation()` and another for `testCustomNonnullAnnotation()`, just to keep each test case specific to one thing.

p.s. Given that separation, it might also be simpler in the `testCustomNonnullAnnotation` to just use a method of an unannotated class that takes `@NoNull` as an argument and pass `null` to it from annotated code, since that's easier to understand at a glance than the override/inheritance logic here.

Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java:3230]: I tested it with the line ```"-XepOpt:NullAway:CustomNonnullAnnotations=qual.NoNull"``` commented and it did fail.

Also separated both tests with simpler logic: e2d4c4c7584ce1665d9592e4a75c2df75ca585a9

**Comment by nimakarimipour** [comment]
@lazaroclapp @msridhar I believe this is ready for another review.

**Review by lazaroclapp — APPROVED** [review]
LGTM. Let me change the description a tiny bit and land this. I will run the performance testing again on a devpod before releasing the next version, just in case (and also to contrast the full current version with the previous release).

**Comment by lazaroclapp** [comment]
Follow up:

Updated our wiki/docs based on this PR.

Docs for the specific new command-line options here: https://github.com/uber/NullAway/wiki/Configuration#custom-nullability-annotations

Also, a revamped discussion of supported nullability annotations here: https://github.com/uber/NullAway/wiki/Supported-Annotations#nullability
